connectors/sources/graphql.py

```diff
@@ -0,0 +1,449 @@
+            raise ConfigurableFieldValueError(msg)
+
+        if not self.is_query(graphql_query=self.graphql_client.graphql_query):
+            msg = "Configured Query is not supported by the connector."
+            raise ConfigurableFieldValueError(msg)
+
+        headers = self.graphql_client.configuration["headers"]
+        graphql_variables = self.graphql_client.configuration["graphql_variables"]
+        if headers:
+            try:
+                self.graphql_client.headers = json.loads(headers)
+            except Exception as exception:
+                msg = f"Error while processing configured GraphQL headers. Exception: {exception}"
+                raise ConfigurableFieldValueError(msg) from exception
+
+        if graphql_variables:
+            try:
+                self.graphql_client.variables = json.loads(graphql_variables)
+            except Exception as exception:
+                msg = f"Error while processing configured GraphQL variables. Exception: {exception}"
+                raise ConfigurableFieldValueError(msg) from exception
+
+    async def close(self):
+        await self.graphql_client.close()
+
+    async def ping(self):
+        try:
+            await self.graphql_client.ping()
+            self._logger.debug("Successfully connected to GraphQL Instance.")
+        except Exception:
+            self._logger.exception("Error while connecting to GraphQL Instance.")
+            raise
+
+    async def fetch_data(self, graphql_query):
+        if self.graphql_client.pagination_enabled:
+            async for data in self.graphql_client.paginated_call(
+                graphql_query=graphql_query
+            ):
+                if isinstance(data, dict):
+                    yield data
+                elif isinstance(data, list):
+                    for document in data:  # pyright: ignore
+                        if isinstance(document, dict):
+                            yield document
+        else:
+            data = await self.graphql_client.make_request(graphql_query=graphql_query)
+            for documents in self.graphql_client.extract_graphql_data_items(
+                key="data", data=data
+            ):
+                if isinstance(documents, dict):
+                    yield documents
+                elif isinstance(documents, list):
+                    for document in documents:
+                        if isinstance(document, dict):
+                            yield document
+
+    async def get_docs(self, filtering=None):
+        """Executes the logic to fetch GraphQL response in async manner.
+
+        Args:
+            filtering (filtering, None): Filtering Rules. Defaults to None.
+
+        Yields:
+            dict: Documents from GraphQL.
+        """
+        doc_id = 1
+        async for doc in self.fetch_data(
+            graphql_query=self.graphql_client.graphql_query
+        ):
+            doc = deepcopy(doc)
+            doc["_id"] = hash_id(str(doc_id))
```

Comment:
I don't think we should autogenerate doc ids. This doesn't make much sense to me. We can easily end up in a situation that a same doc has assigned different ids in different full syncs. 

This breaks assumption in the framework that docs have unique ids that don't change. Could we perhaps utilize [graphql global object ids](https://graphql.org/learn/global-object-identification/)?

Reply:
After telling users to add unique ID how can we identify that unique field ? As every source can have a unique name for ID (id, uid, sys_id etc.) different from the other source, we would not be able to tell which is the unique field. Also in this connector I guess it is fine if same id assigned to different document as we are indexing all documents in every sync so previous document will get new ID.

As most of the source have unique field as `id`. So, for now i am using that id in case if it present in the doc otherwise uses custom id